[RFC] Lua error message better formatting

AlexandreDubray opened the pull request:
reference #7388

Commit:
message: moving emsg to _emsg

_emsg is the same as emsg but has additionnal argument which is a pointer
to a function that is used to return the value

ZyX-I commented on src/nvim/message.c:
```diff
@@ -572,8 +564,22 @@ int emsg(const char_u *s_)
+ *
+ * return TRUE if wait_return not called
+ */
+int emsg(const char_u *s_)
```
No need to have `const char *s = (const char *)s_;` line above as well as underscore here: underscore is only because of the existence of the line and line is there only because I did not feel like refactoring lots of `emsg` callers which use `char_u`.

ZyX-I replied:
Ah, and here still is underscore.

AlexandreDubray replied:
Oh god, I must be blind. Sorry for that, I will change it immediately so I will not forget anymore !

ZyX-I commented on src/nvim/message.c:
```diff
@@ -466,15 +466,7 @@ int emsg_not_now(void)
- * return TRUE if wait_return not called
- */
-int emsg(const char_u *s_)
+static int _emsg(const char *s_, int (*msg_attr)(const char *s, const int attr))
```
It is bad idea to mask symbols; also need typedef, AFAIR C declarations generator intentionally cannot swallow that.

AlexandreDubray replied:
I made the typedef in a commit but I did not understood the "mask symbol" part (I mean, technically what does that mean? I'm sorry, my knowledge of C is not that advanced).

ZyX-I replied:
It means that you should not define a symbol in some inner scope which is the same as a symbol in an outer scope. E.g. in

```C
int i = 0;
{
  int i = 1;
  printf("%d\n", i);
}
printf("%d\n", i);
```

second `i` masks the first one. `msg_attr` is a function defined in compilation unit scope, and at the same time an argument to `_emsg` which means that it is redefined in `_emsg` functions scope.

AlexandreDubray replied:
Indeed it was not smart, thanks for the explanation!

Commit:
message.c: add msg_echo_attr functions

The added function behaves like the non-echo function but display message
in a echo-style way (i.e. tab and newline are preserved)

Commit:
message.c: add typedef for msg_attr function to be passed to _emsg

Commit:
message.c: remove old cast in _emsg

Commit:
message.c: lint

ZyX-I commented on src/nvim/message.c:
```diff
@@ -54,6 +54,8 @@ struct msgchunk_S {
   char_u sb_text[1];            /* text to be displayed, actually longer */
 };
 
+typedef int (*fct_msg_attr)(const char *s, const int attr);
```
New types should be CamelCase.

ZyX-I commented on src/nvim/message.c:
```diff
@@ -137,6 +139,94 @@ int msg_attr(const char *s, const int attr) FUNC_ATTR_NONNULL_ARG(1)
+msg_echo_attr_keep(
+    const char *s,
+    const int attr,
+    int keep
```
You must not put arguments completely on separate lines unless you absolutely have to. Neither separate type from function name or closing parenthesis from the last argument.

ZyX-I commented on src/nvim/message.c:
```diff
@@ -137,6 +139,94 @@ int msg_attr(const char *s, const int attr) FUNC_ATTR_NONNULL_ARG(1)
+    const int attr,
+    int keep
+)
+  FUNC_ATTR_NONNULL_ARG(1)
```
FUNC_ATTR_NONNULL_ALL may be used when all *pointer* arguments are non-NULL. Non-pointer ones do not matter.

ZyX-I commented on src/nvim/message.c:
```diff
@@ -577,9 +666,9 @@ static int _emsg(const char *s_, int (*msg_attr)(const char *s, const int attr))
 int emsg(const char_u *s_)
 {
-  return _emsg((const char *)s_, &msg_attr);
+  fct_msg_attr f = &msg_attr;
```
Do not create temporary variable here, code should work as it worked.

ZyX-I commented on src/nvim/message.c:
```diff
@@ -137,6 +139,94 @@ int msg_attr(const char *s, const int attr) FUNC_ATTR_NONNULL_ARG(1)
+  buf = (char *)msg_strtrunc((char_u *)s, false);
+  if (buf != NULL) {
+    s = buf;
+  }
```
Too much copy-paste.

ZyX-I commented on src/nvim/message.c:
```diff
@@ -137,6 +139,94 @@ int msg_attr(const char *s, const int attr) FUNC_ATTR_NONNULL_ARG(1)
+
+  char * spec_char = "\t\n\r";
+
+  char * next_spec = (char *)s;
```
There should not be space after asterisk.

ZyX-I replied:
And `next_spec` must be `const char *`. Do not cast to `char *` until needed.

ZyX-I replied:
I.e. until going to legacy (unrefactored) msg_outtrans[_len]_attr.

ZyX-I commented on src/nvim/message.c:
```diff
@@ -137,6 +139,94 @@ int msg_attr(const char *s, const int attr) FUNC_ATTR_NONNULL_ARG(1)
+    s = buf;
+  }
+
+  char * spec_char = "\t\n\r";
```
Do not make compiler work harder, strpbrk argument better be a literal, either just literal or macros. But since you are using this list in a single place just put literal there.

ZyX-I commented on src/nvim/message.c:
```diff
@@ -137,6 +139,94 @@ int msg_attr(const char *s, const int attr) FUNC_ATTR_NONNULL_ARG(1)
+  char * next_spec = (char *)s;
+
+  while (next_spec != NULL) {
+    next_spec = (char *)vim_strpbrk(s, spec_char);
```
You just added a bunch of useless casts there (some hidden by `vim_strpbrk` macros), don’t use `vim_strpbrk`, use `strpbrk` directly.

ZyX-I commented on src/nvim/message.c:
```diff
@@ -606,6 +713,19 @@ static bool emsgfv(const char *fmt, va_list ap)
   return emsg((const char_u *)errbuf);
 }
 
+static bool emsgfv_echo(const char *fmt, va_list ap)
```
This function is short and not used anywhere, but in emsgf_echo, so should not exist.

ZyX-I commented on src/nvim/message.c:
```diff
@@ -572,7 +653,21 @@ int emsg(const char_u *s_)
   // Display the error message itself.
   msg_nowait = false;  // Wait for this msg.
-  return msg_attr(s, attr);
+  return ret_fct(s, attr);
```
`ret_fct` is a strange name. I would go with something like `display_msg`.

Commit:
message.c: style

Commit:
message.c: remove useless call to add_msg_hist in msg_echo_attr_keep

Commit:
message.c: msg_echo style + remove useless variable

Commit:
style

AlexandreDubray commented:
@ZyX-I I think I've modified everything you pointed out. Still I'm not sure about the "too much copy-paste" comment. Let me know if I need to modify anything :-)

ZyX-I commented on src/nvim/message.c:
```diff
@@ -137,6 +139,81 @@ int msg_attr(const char *s, const int attr) FUNC_ATTR_NONNULL_ARG(1)
+}
+
+int
+msg_echo_attr_keep(const char *s, const int attr, int keep)
```
You are still separating type from function name. Correct examples of formatting are new files like src/nvim/eval/typval.c, not legacy code here.

ZyX-I replied:
BTW, function returns boolean.

ZyX-I replied:
And it is logical to either have `const int keep` or `int attr`. I normally prefer putting `const` everywhere I can.

ZyX-I commented on src/nvim/message.c:
```diff
@@ -137,6 +139,81 @@ int msg_attr(const char *s, const int attr) FUNC_ATTR_NONNULL_ARG(1)
   return msg_attr_keep((char_u *)s, attr, false);
 }
 
+int msg_echo_attr(const char *s, const int attr) FUNC_ATTR_NONNULL_ARG(1)
```
Do not put FUNC_ATTR on the same line as a function name, this variant both visually hides attributes and produces unneccessary line change if something needs to be added (e.g. new argument or another attribute).

ZyX-I replied:
BTW, is this a function intended to be used in `ex_echo`? It is not good to have two implementations doing essentially the same task; also using function there makes testing a lot easier and allows CI to catch errors before test suite is good enough.

AlexandreDubray replied:
I did not intend to use it in `ex_echo`, but I guess I could modify `ex_echo` to use it (if it is the purpose of the question)

ZyX-I commented on src/nvim/message.c:
```diff
@@ -137,6 +139,81 @@ int msg_attr(const char *s, const int attr) FUNC_ATTR_NONNULL_ARG(1)
+msg_echo_attr_keep(const char *s, const int attr, int keep)
+  FUNC_ATTR_NONNULL_ALL
+{
+  static int entered = 0;
```
This function contains too many copy-paste from its origin, common code should not be copied if possible.

ZyX-I commented on src/nvim/message.c:
```diff
@@ -137,6 +139,81 @@ int msg_attr(const char *s, const int attr) FUNC_ATTR_NONNULL_ARG(1)
+{
+  static int entered = 0;
+  int retval;
+  char * buf = NULL;
```
Here `*` asterisk is formatted as multiplication, but it is not.

ZyX-I commented on src/nvim/message.c:
```diff
@@ -137,6 +139,81 @@ int msg_attr(const char *s, const int attr) FUNC_ATTR_NONNULL_ARG(1)
+  }
+
+  msg_start();
+  buf = (char *)msg_strtrunc((char_u *)s, false);
```
And if you do refactor old code, move variable definitions closer to the place where they are actually used. It should be `char *const buf` here.

ZyX-I commented on src/nvim/message.c:
```diff
@@ -137,6 +139,81 @@ int msg_attr(const char *s, const int attr) FUNC_ATTR_NONNULL_ARG(1)
+    s = buf;
+  }
+
+  const char *spec_char = "\t\n\r";
```
This variable still has no reason to exist.

AlexandreDubray replied:
Oh I see, I misunderstood your previous comment

ZyX-I commented on src/nvim/message.c:
```diff
@@ -137,6 +139,81 @@ int msg_attr(const char *s, const int attr) FUNC_ATTR_NONNULL_ARG(1)
+  FUNC_ATTR_NONNULL_ALL
+{
+  static int entered = 0;
+  int retval;
```
This one should not exist either, need `const bool retval` below.

ZyX-I commented on src/nvim/message.c:
```diff
@@ -137,6 +139,81 @@ int msg_attr(const char *s, const int attr) FUNC_ATTR_NONNULL_ARG(1)
+  retval = msg_end();
+
+  if (keep && retval && vim_strsize((char_u *)s) < (int)(Rows - cmdline_row -1)
+      * Columns + sc_col) {
```
Alignment is not correct.

Commit:
message.c: style and convention

AlexandreDubray commented on src/nvim/message.c:
```diff
@@ -134,24 +134,23 @@ int verb_msg(char_u *s)
 
-int
-msg_echo_attr_keep(const char *s, const int attr, int keep)
+bool msg_attr_skip(const char *s, const int attr, int *const call)
```
@ZyX-I I move the common code into this function that return when the message doesn't need to be shown (I will add comments, I forgot them). Is the function ok for you? 
Also, i didn't put the `add_msg_hist` here because (if I understood well the code) it is already done in `msg_outtrans_len_attr` and `msg_puts_attr_len`. I guess it doesn't need to be done multiple time?

justinmk commented on src/nvim/message.c:
```diff
@@ -572,7 +641,20 @@ int emsg(const char_u *s_)
+ * When terminal not initialized (yet) mch_errmsg(..) is used.
+ *
+ * return TRUE if wait_return not called
+ */
```
Please run `make lint`, it will notify you about lint problems.

justinmk commented:
typedef int (*FctMsgAttr)(const char *s, const int attr);

What is `Fct`? If it means "function", that's very unusual. `Fn` or `Fun` or `Func` are far more common. Use `git grep` to see existing examples:

    git grep -E 'typedef.*\(.*\)' -- ':/*.c'

Commit:
message.c: lint and style

AlexandreDubray commented:
@justinmk Yep `Fct` was for function (old habits die hard ^^) corrected now! (plus the lint problem). I think it's ready if there is nothing else :-)

ZyX-I commented:
I do not see new code used for echo. Also missing tests including outputting following messages:

1. Regular ones with multiline stack trace.
2. Single-line regular (i.e. what a sane developer would want to throw with `error()`) ones.
3. Consisting of just three newlines, three tabs and three `\r`s and nothing else. (These are *three* cases, not one).
4. Long (longer then screen width) single-line ones with tabs inside.
5. Multiline with long (…) lines.
6. Just empty string.

That list repeated for

1. Throwing lua errors and displaying them on screen.
2. Throwing lua errors and catching them with `redir`, `execute()` or `nvim_command_output` (since it is easier to write test generator then repeat tests it would make sense to use all variants).
3. Throwing lua errors and catching them with `:try|catch`.
4. Echoing messages with `:echo` and displaying them on screen.
5. Echoing messages with `:echon` and displaying them on screen.
6. Echoing messages with `:echo` and catching them with `redir`/… (…).
7. Echoing messages with `:echon` and catching them with … (…).

As I said I would expect you to write a test generator rather then lots of boilerplate tests.

AlexandreDubray commented:
Oh yeah, I forget this story about the `echo`, will do that. And also, what do you mean exactly by test generator ?

ZyX-I commented:
Normally you write tests like `describe('text', function() it('text', function() --[[ actual test ]] end) end)`, placing that directly in the file. By “test generator” I mean some lua function which calls `describe()` and/or `it()` with computed text and actual test being a closure acting based on test generator arguments. It is used throught the test code for creating similar tests: e.g. in https://github.com/ZyX-I/neovim/blob/5c84d48b1f0ba2400ffb06f20ceb2da3e0590261/test/functional/eval/null_spec.lua or https://github.com/ZyX-I/neovim/blob/5c84d48b1f0ba2400ffb06f20ceb2da3e0590261/test/functional/ex_cmds/append_spec.lua.

Commit:
message.c: style

AlexandreDubray commented:
Ok, thanks for the references, I will look at that.

Commit:
ex_echo: reuse code from message.c to show arg to user

justinmk commented:
Merged by #9520. Would be good if someone implemented the tests mentioned in https://github.com/neovim/neovim/pull/7969#issuecomment-365427602 .
